src/SurfaceFluxes.jl

```diff
@@ -37,436 +42,717 @@ include("UniversalFunctions.jl")
+  - L_MO:   Monin-Obukhov lengthscale
+  - shf:    Sensible Heat Flux
+  - lhf:    Latent Heat Flux
+  - ρτxz:   Momentum Flux (Eastward component)
+  - ρτyz:   Momentum Flux (Northward component)
+  - ustar:  Friction velocity
+  - Cd:     Momentum Exchange Coefficient
+  - Ch:     Thermal Exchange Coefficient
 """
 function surface_conditions(
     param_set::APS,
-    MO_param_guess::AbstractVector,
-    x_in::AbstractVector,
-    x_s::AbstractVector,
-    z_0::Union{AbstractVector, FT},
-    θ_scale::FT,
-    z_in::FT,
-    scheme,
-    wθ_flux_star::Union{Nothing, FT} = nothing,
-    universal_func::Union{Nothing, F} = UF.Businger,
-    sol_type::NS.SolutionType = NS.CompactSolution(),
-    tol::NS.AbstractTolerance = NS.ResidualTolerance{FT}(sqrt(eps(FT))),
-    maxiter::Int = 10_000,
-) where {FT <: AbstractFloat, APS, F}
-
-    n_vars = length(MO_param_guess) - 1
-    @assert length(MO_param_guess) == n_vars + 1
-    @assert length(x_in) == n_vars
-    @assert length(x_s) == n_vars
-    local sol
-
-    args = (;
-        param_set,
-        wθ_flux_star,
-        z_in,
-        z_0,
-        n_vars,
-        x_in,
-        x_s,
-        scheme,
-        θ_scale,
-        MO_param_guess,
-        universal_func,
-    )
-
-    # Define closure over args
-    f!(F, x_all) = surface_fluxes_f!(F, x_all, args)
+    sc::AbstractSurfaceConditions,
+    universal_func = UF.Businger,
+    scheme::SolverScheme = FVScheme(),
+)
+    # Compute output variables for SurfaceFluxConditions
+    FT = eltype(sc.state_in.u)
+    L_MO = obukhov_length(param_set, sc, universal_func, scheme)
+    ustar = compute_ustar(param_set, L_MO, sc, universal_func, scheme)
+    Cd = momentum_exchange_coefficient(param_set, L_MO, sc, universal_func, scheme)
+    Ch = heat_exchange_coefficient(param_set, L_MO, sc, universal_func, scheme)
+    shf = sensible_heat_flux(param_set, Ch, sc, scheme)
+    lhf = latent_heat_flux(param_set, Ch, sc, scheme)
+    buoy_flux = compute_buoyancy_flux(param_set, shf, lhf, ts_in(sc), ts_sfc(sc), scheme)
+    ρτxz, ρτyz = momentum_fluxes(param_set, Cd, sc, scheme)
+    return SurfaceFluxConditions{FT}(L_MO, shf, lhf, buoy_flux, ρτxz, ρτyz, ustar, Cd, Ch)
+end
 
-    nls = NS.NewtonsMethodAD(f!, MO_param_guess)
-    sol = NS.solve!(nls, sol_type, tol, maxiter)
+"""
+    obukhov_length(param_set, sc, universal_func, scheme)
 
-    root_tup = Tuple(sol.root)
+Compute and return the Monin-Obukhov length using non-linear solver iterations.
+"""
+function obukhov_length(param_set, sc::AbstractSurfaceConditions, universal_func, scheme)
+    FT = eltype(sc.L_MO_init)
+    function root_l_mo(x_lmo)
+        root = x_lmo - local_lmo(param_set, x_lmo, sc, universal_func, scheme)
+        return root
+    end
+    sol = RS.find_zero(
+        root_l_mo,
+        RS.NewtonsMethodAD(sc.L_MO_init),
+        RS.CompactSolution(),
+        RS.SolutionTolerance(max(FT(1e-2), sqrt(eps(FT)))),
+        100,
```

Comment:
Should we expose the max iters to the user?

Reply:
@akshaysridhar can you add this?